app/javascript/components/student/Editor.tsx

```diff
@@ -17,39 +18,104 @@ export enum TestRunStatus {
   OPS_ERROR = 'ops_error',
 }
 
+type State = {
+  submission?: Submission
+  isSubmitting: boolean
+  code: string
+}
+
+export type Action =
+  | { type: 'submitting'; payload: { code: string } }
+  | { type: 'submitted'; payload: { submission: Submission } }
+  | { type: 'cancelled' }
+
+function reducer(state: State, action: Action): State {
+  switch (action.type) {
+    case 'submitting':
+      return {
+        submission: undefined,
+        code: action.payload.code,
+        isSubmitting: true,
+      }
+    case 'submitted':
+      return {
+        ...state,
+        submission: action.payload.submission,
+        isSubmitting: false,
+      }
+    case 'cancelled':
+      return {
+        ...state,
+        isSubmitting: false,
+      }
+    default:
+      return state
+  }
+}
+
 export function Editor({
   endpoint,
   timeout = 60000,
 }: {
   endpoint: string
   timeout?: number
 }) {
-  const [submission, setSubmission] = useState<Submission>()
+  const [{ submission, isSubmitting, code }, dispatch] = useReducer(reducer, {
+    submission: undefined,
+    isSubmitting: false,
+    code: '',
+  })
   const controllerRef = useRef(new AbortController())
+  const abortFetch = useCallback(() => {
+    controllerRef.current.abort()
+    controllerRef.current = new AbortController()
+  }, [controllerRef.current])
 
   useEffect(() => {
     return () => {
-      controllerRef.current.abort()
+      abortFetch()
     }
   }, [])
 
-  function submit(code: string) {
-    controllerRef.current.abort()
-    controllerRef.current = new AbortController()
-    setSubmission(undefined)
+  useEffect(() => {
```

Comment:
I find it hard to follow why we need all the different effects. So apparently `isSubmitting` can change on its own. And then you want to abort? Why? It feels like you only want that because you want a `new AbortController`.